Structural/adjoint fields

MFusseder opened the pull request:
Within this PR I added the particular solution of the influence function of stress resultants for beam and truss elements. The PR also includes the attempt to post-process curvature and strain of the adjoint displacement.

MFusseder committed:
add variable for adjoint particular displacement

MFusseder committed:
add possibility to compute post processing quantities of adjoint field

MFusseder committed:
don't call finalize solution step of response in scheme

MFusseder committed:
add particular solution for inflfunc of beam stress resultants

MFusseder committed:
add variables for adjoint field quantities of beam

MFusseder committed:
compute adjoint curvature & strain within adjoint beam element

MFusseder committed:
add warning in case Ay or Az are unequal to zero

MFusseder committed:
transform particular solution in global direction

MFusseder committed:
perform transformation within adjoint beam element

MFusseder committed:
remove useless include

MFusseder committed:
move adjoint field computation to header

MFusseder committed:
added adjoint field computation for truss element

MFusseder committed:
code improvement

MFusseder committed:
use same functions for beam and truss

MFusseder committed:
clean up

MFusseder committed:
add warning concerning OMP

MFusseder committed:
Merge branch 'master' into structural/adjoint-fields

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp:
```diff
@@ -28,6 +29,76 @@ AdjointFiniteDifferenceCrBeamElement::~AdjointFiniteDifferenceCrBeamElement()
+
+        KRATOS_ERROR_IF(rOutput.size() != element_size) << "Size of particular solution does not fit!" << std::endl;
+
+        CrBeamElement3D2N::Pointer p_primal_beam_element = dynamic_pointer_cast<CrBeamElement3D2N>(this->pGetPrimalElement());
```
please do not do casts, refactor the existing functions instead (I recently started moving common functionalities to strucutral_mechancis_element_utilities)

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp:
```diff
@@ -28,6 +29,76 @@ AdjointFiniteDifferenceCrBeamElement::~AdjointFiniteDifferenceCrBeamElement()
+        {
+            this->CalculateAdjointFieldOnIntegrationPoints(FORCE, rOutput, rCurrentProcessInfo);
+
+            KRATOS_WARNING_IF("ADJOINT_STRAIN", (this->GetProperties().Has(AREA_EFFECTIVE_Y) || this->GetProperties().Has(AREA_EFFECTIVE_Z)))
```
usually we just ignore additionally defined variables

MFusseder replied:
In this special case I like to keep the warning

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_linear_3D2N.cpp:
```diff
@@ -27,6 +28,53 @@ AdjointFiniteDifferenceTrussElementLinear::~AdjointFiniteDifferenceTrussElementL
+
+        KRATOS_ERROR_IF(rOutput.size() != element_size) << "Size of particular solution does not fit!" << std::endl;
+
+        TrussElement3D2N::Pointer p_primal_beam_element = dynamic_pointer_cast<TrussElement3D2N>(this->pGetPrimalElement());
```
same

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -327,5 +342,178 @@ namespace Kratos
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::CalculateNodeParticularSolutionLinearElement2N(Vector& rResult, DofsVectorType &rElementalDofList, const std::string& rDofLabel)
```
const

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -327,5 +342,178 @@ namespace Kratos
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::FindCorrespondingDofLabel(std::string& rDofLabel)
```
const

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.h:
```diff
@@ -197,6 +203,18 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) AdjointLocalStressResponseFun
+
+    void CalculateNodeParticularSolutionLinearElement2N(Vector& rResult, DofsVectorType &rElementalDofList, const std::string& rDofLabel);
+
+    void FindCorrespondingDofLabel(std::string& rDofLabel);
```
I think most of those functions can be `const` since they don't change members

philbucher reviewed (changes requested):
I briefly looked at this and saw some things
- please remove the tetgen-module
- do not do the casts, rather refactor the functions that are necessary
- make functions that do not modify (or should not modify) members of a class `const`

MFusseder committed:
delete tetgen

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h:
```diff
@@ -400,6 +417,77 @@ class AdjointFiniteDifferencingBaseElement : public Element
+
+        // Build vector of variables containing the DOF-variables of the primal problem
+        std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list;
+        primal_solution_variable_list.reserve(num_dofs_per_node);
```
why are you storing the components independently and not `DISPLACEMENT` directly?
Same for `ROTATION` below

armingeiser replied:
Ok i think i got the reason.

MFusseder committed:
change functions to const

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp:
```diff
@@ -28,6 +29,76 @@ AdjointFiniteDifferenceCrBeamElement::~AdjointFiniteDifferenceCrBeamElement()
+        }
+    }
+    else
+        this->CalculateAdjointFieldOnIntegrationPoints(rVariable, rOutput, rCurrentProcessInfo);
```
Is `CalculateAdjointFieldOnIntegrationPoints` a valid function for any variable except `ADJOINT_CURVATURE` and `ADJOINT_STRAIN`

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -37,10 +39,23 @@ namespace Kratos
         }
+
+        if(ResponseSettings.Has("consider_particular_solution"))
+            mAddParticularSolution = ResponseSettings["consider_particular_solution"].GetBool();
```
Just for convenience: can you use similar names for the json setting and the variable in cpp?

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -327,5 +342,178 @@ namespace Kratos
+        if(element_name == "CrLinearBeamElement3D2N" || element_name == "TrussLinearElement3D2N")
+        {
+            Vector particular_solution;
+            // delifers particular solution of influence function in local coordinates
```
delivers

MFusseder committed:
some renaming according to review

MFusseder commented:
Thank you for the review. I addressed your all your suggestions for improvement. The only remaining question is how to prevent the casts. Hopefully I can clarify this topic the next days.

MFusseder committed:
add finalizesolutionstep again

MFusseder committed:
resolve conflicts and update branch for changes in master

MFusseder committed:
temporal test how to handle primal element

MFusseder committed:
Merge branch 'master' into structural/adjoint-fields

philbucher commented:
ping is this still active?

MFusseder commented:
Yes it is. There is still the open question on how to prevent the casts. Hopefully I can continue to work on this topic soon.

philbucher commented:
is this still active?

MFusseder commented:
I am still interested to have this in master. But I am not sure when I will find time to continue with it.

MFusseder committed:
erase temporal test how to handle primal element again

MFusseder committed:
solve conflicts after merging master into branch

MFusseder committed:
modified transformation of particular solution

MFusseder commented:
I finally get rid of the casts. Now I receive the transformation with the help of `Calculate(...)` when calling it with the variable `LOCAL_ELEMENT_ORIENTATION` similar to the shell elements as @philbucher suggested. Sorry for the long downtime of this merge request.

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1188,6 +1188,15 @@ CrBeamElement3D2N::GetCurrentNodalPosition() const
+{
+    if (rVariable == LOCAL_ELEMENT_ORIENTATION)
+    {
+        rOutput.resize(msElementSize, msElementSize, false);
```
resizing should always be preceded with a check to see if it is actually necessay

please also change this in the other places

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1188,6 +1188,15 @@ CrBeamElement3D2N::GetCurrentNodalPosition() const
+    if (rVariable == LOCAL_ELEMENT_ORIENTATION)
+    {
+        rOutput.resize(msElementSize, msElementSize, false);
+        noalias(rOutput) = GetTransformationMatrixGlobal();
```
@KlausBSautter we should speak how to make this more efficient => the fct `GetTransformationMatrixGlobal` should get the matrix as an input and not return it

MFusseder committed:
add size check before resizing of output matrix

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h:
```diff
@@ -434,6 +451,77 @@ class AdjointFiniteDifferencingBaseElement : public Element
+            particular_solution = this->GetValue(ADJOINT_PARTICULAR_DISPLACEMENT);
+
+        // Build vector of variables containing the DOF-variables of the primal problem
+        std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list;
```
```suggestion
        std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>*> primal_solution_variable_list;
```
I recommend using pointers here, there might be problems with copying variables

MFusseder committed:
use pointers within primal & adjoint variables list

philbucher reviewed (dismissed):
ok for me

MFusseder commented:
great! thank you for your help :-)

armingeiser commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -258,6 +258,11 @@ class CrBeamElement3D2N : public Element
     BoundedVector<double,msElementSize> CalculateBodyForces() const;
 
+    void Calculate(
+            const Variable<Matrix>& rVariable,
```
indentation

armingeiser commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.hpp:
```diff
@@ -95,6 +95,11 @@ class CrBeamElementLinear3D2N : public CrBeamElement3D2N
     BoundedMatrix<double,msLocalSize,msLocalSize> CalculateDeformationStiffness() const override;
 
+    void Calculate(
+            const Variable<Matrix>& rVariable,
```
indentation

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h:
```diff
@@ -434,6 +451,77 @@ class AdjointFiniteDifferencingBaseElement : public Element
+        // Build vector of variables containing the DOF-variables of the primal problem
+        std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>*> primal_solution_variable_list;
+        primal_solution_variable_list.reserve(num_dofs_per_node);
+        primal_solution_variable_list.push_back(&DISPLACEMENT_X);
```
i think using initilizerlist and tenary operator this could be written with much less lines:

```cpp
typedef std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>*> variable_pointer_vector;
const variable_pointer_vector primal_solution_variable_list = mHasRotationDofs ? 
    {&DISPLACEMENT_X, &DISPLACEMENT_Y, &DISPLACEMENT_Z, &ROTATION_X, &ROTATION_Y, &ROTATION_Z} 
    : {&DISPLACEMENT_X, &DISPLACEMENT_Y, &DISPLACEMENT_Z};
```

up to you if you like it.

philbucher replied:
That looks nice :)

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -328,5 +343,185 @@ namespace Kratos
+        {
+            case TracedStressType::MX:
+            {
+                rDofLabel = "ADJOINT_ROTATION_X";
```
I have the feeling that passing a variable instead of a string would be more efficient here. up to the point where you compare with the dof variable in CalculateNodeParticularSolutionLinearElement2N

armingeiser replied:
then the name for this function would be FindVariableComponent or similar

armingeiser replied:
In that case a typdef for variable component would be good i guess =)

armingeiser reviewed (commented):
I have some comments, none of them is critical but just let me know what you think about it.

MFusseder committed:
indentation cosmetic

MFusseder committed:
use variable component instead of string to handle corresponding dof

MFusseder committed:
call also basetype within cnew calculate()

MFusseder committed:
enhance declaration of dof lists

MFusseder committed:
cosmetics concerning the brackets of if and for constructions

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp:
```diff
@@ -430,90 +424,90 @@ namespace Kratos
     }
 
-    void AdjointLocalStressResponseFunction::FindCorrespondingDofLabel(std::string& rDofLabel) const
+    void AdjointLocalStressResponseFunction::FindVariableComponent(Array1DComponentsPointerType& rTracedDof) const
```
Ref of a pointer?

MFusseder replied:
this was also new to me. @philbucher suggested doing so.

armingeiser replied:
Ok got it, but i think it is only necessary to pass it by reference in the FindVariableComponent function, the otheres are just using it and it could be passed by value there.

philbucher replied:
yes for the other fct it should be sufficent to only pass the pointer (without reference), since there you don't want to change the pointer (only use ti)

MFusseder replied:
thanks for the hint. I pass now the directly the pointer for the other fcts.

MFusseder committed:
pass TracedDof of by value instead of by reference

MFusseder committed:
Merge branch 'master' into structural/adjoint-fields

MFusseder commented:
@armingeiser and @philbucher are the modifications now fine for you?

philbucher commented:
Fine for me
